lib/OpenQA/Worker/Job.pm

```diff
@@ -781,30 +781,29 @@ sub _upload_results_step_0_prepare {
     if ($self->{_has_uploaded_logs_and_assets} || $running_or_finished) {
         my @file_info = stat $self->_result_file_path('test_order.json');
         my $test_order;
-        if (   !$current_test_module
-            or !$file_info[9]
-            or $file_info[9] != $self->{_test_order_mtime}
-            or !$file_info[7]
-            or $file_info[7] != $self->{_test_order_fsize})
-        {
-            log_info('Test schedule has changed, reloading test_order.json') if $self->{_test_order_mtime};
-            $test_order                = $self->_read_json_file('test_order.json');
-            $status{test_order}        = $test_order;
-            $self->{_test_order}       = $test_order;
-            $self->{_test_order_mtime} = $file_info[9];
-            $self->{_test_order_fsize} = $file_info[7];
-        }
-        if (!$current_test_module) {    # first test (or already after the last!)
+        if (!exists $self->{_test_order_mtime}) {    # fist read test_order.json
+            $test_order = $self->_read_json_file('test_order.json');
             if (!$test_order) {
-                # FIXME: It would still make sense to upload other files.
                 $self->stop('no tests scheduled');    # will be delayed until upload has been concluded
                 $self->emit(uploading_results_concluded => {});
                 return Mojo::IOLoop->next_tick($callback);
             }
+            $status{test_order}        = $test_order;
+            $self->{_test_order}       = $test_order;
+            $self->{_test_order_mtime} = $file_info[9];
+            $self->{_test_order_fsize} = $file_info[7];
         }
-        elsif ($current_test_module ne $running_test) {    # next test
-            $upload_up_to = $current_test_module;
+        else {
+            if ($file_info[9] != $self->{_test_order_mtime} || $file_info[7] != $self->{_test_order_fsize}) {
+                $test_order = $self->_read_json_file('test_order.json');
+                log_info('Test schedule has changed, reloading test_order.json');
+                $status{test_order}        = $test_order;
+                $self->{_test_order}       = $test_order;
+                $self->{_test_order_mtime} = $file_info[9];
+                $self->{_test_order_fsize} = $file_info[7];
```

Comment:
This is getting repetitive. Maybe it is worth creating a separate function, at least for the last three assignments, e.g. `$self->_assign_test_order($test_order, \@file_info)`.